jobserver/hash_utils.py

```diff
@@ -90,6 +91,16 @@ def hash(m, length=DEFAULT_LENGTH, key=DEFAULT_KEY):  # noqa: A001
     return hex(m_hash)[2:].rjust(length, "0")
 
 
+def hash_user_pat(token):
+    """
+    Utility function to hash a token for User PATs
+
+    The hashingn algorithm for this is controlled by the PASSWORD_HASHERS
+    setting.
+    """
+    return make_password(token, salt="user_pat")
```

Comment:
The docs don't really suggest what the salt is used for.  The code uses whatever the hasher has set (for Argon2 this is the default value of the base hasher).  I've set this salt as a way to separate hashes generated by this function from other uses of the function, eg if we were to store passwords.

Reply:
Using a constant value for the salt means that you'll get the same hash back each time it's called for the same password (or token in this case). I think by default a random salt is used, which results in a different hash each time. It protects against attacks using rainbow tables.

I would be tempted to go with the default behaviour of random salts, unless we need that consistency?

Reply:
I feel like I'm missing something fundamental here but surely we want the same hash back every time we hash a given string, how else do we compare them?

Reply:
The salt is stored alongside the hash, so will be used by check_password. There's a pretty good write-up on [wikipedia](https://en.wikipedia.org/wiki/Salt_(cryptography)) about how and why salts are used.

Reply:
Salt have a broader usage than just password hashing.

For user-supplied passwords, which are often english words (other languages are available), are vulnerable to pre-calculated dictionary (rainbow tables). So such systems typically include a per-user salt, so that you cannot use the same rainbow table to attack all the hashes, you'd need to do it per user.

Here however, we are not using user supplied secrets - only we are ever setting the secret, and it has high entropy, so a rainbow attack is not feasible anyway. Thus, fix salt is fine.

Fixed salts like this are commonly used when signing things, like we do in [release-hatch](https://github.com/opensafely-core/release-hatch/blob/main/hatch/signing.py#L21).  It partitions the signature space by salt, to limit cross-service replay attacks. In this way, it's actually more like a [pepper](https://en.wikipedia.org/wiki/Pepper_(cryptography)) than a salt (but it is not a pepper currently, although we could make it so).

It's particularly confusing here, because this django implementation of salted hashes is totally intended for user supplied passwords.

Reply:
Thank you for the explanation, that makes more sense now. I was definitely thinking more along the lines of the approach taken for user-supplied passwords

Reply:
Thank you both for the explanation!